Import Pandas and Numpy.

In [2]:
import pandas as pd
import numpy as np

import warnings
warnings.simplefilter('ignore', FutureWarning)

C:\Users\hp\AppData\Local\Temp\ipykernel_18632\1505564052.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Download the data and read it into a Pandas dataframe.

In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

Check the number of the data points.

In [4]:
concrete_data.shape

(1030, 9)

Check the statistical summary of the data.

In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

Check the data for any missing values.

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

Split data into target (the column 'Strength') and predictors (the other columns).

In [7]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [23]:
predictors.head() # Check the first five rows of the Predictors dataframe

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [24]:
target.head() #Chech the first five values of the Strength column 

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

## Normalize the data by substracting the mean and dividing by the standard deviation.

In [11]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Save the number of predictors to *n_cols*.

In [12]:
n_cols = predictors_norm.shape[1] # number of predictors

Import Keras and its required packages.

In [13]:
import keras

In [14]:
from keras.models import Sequential
from keras.layers import Dense

## Build a Neural Network

In [15]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model with 1 hidden layer and 10 hidden units.


## Train and Test the Network

In [16]:
# build the model
model = regression_model()

C:\Users\hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [17]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=2)

Epoch 1/50
23/23 - 15s - 638ms/step - loss: 1639.8152 - val_loss: 1195.3136
Epoch 2/50
23/23 - 0s - 12ms/step - loss: 1619.6119 - val_loss: 1183.6796
Epoch 3/50
23/23 - 0s - 13ms/step - loss: 1598.7471 - val_loss: 1171.6350
Epoch 4/50
23/23 - 0s - 17ms/step - loss: 1577.0168 - val_loss: 1159.1195
Epoch 5/50
23/23 - 0s - 12ms/step - loss: 1554.5812 - val_loss: 1146.0953
Epoch 6/50
23/23 - 0s - 12ms/step - loss: 1531.2966 - val_loss: 1132.2524
Epoch 7/50
23/23 - 0s - 11ms/step - loss: 1506.1497 - val_loss: 1117.8569
Epoch 8/50
23/23 - 0s - 12ms/step - loss: 1479.5409 - val_loss: 1102.6875
Epoch 9/50
23/23 - 0s - 11ms/step - loss: 1451.2800 - val_loss: 1086.7003
Epoch 10/50
23/23 - 0s - 12ms/step - loss: 1421.4436 - val_loss: 1069.9075
Epoch 11/50
23/23 - 0s - 11ms/step - loss: 1389.5940 - val_loss: 1052.2209
Epoch 12/50
23/23 - 0s - 11ms/step - loss: 1356.1895 - val_loss: 1033.6606
Epoch 13/50
23/23 - 0s - 10ms/step - loss: 1321.6782 - val_loss: 1014.3387
Epoch 14/50
23/23 - 0s - 12ms/st

Repeat the traiing and testing for 50 cycles and create a list of 50 mean squared errors representing each cycle.

In [20]:
mean_squared_errors = []
for i in range(50):
    #Randomly split the data (70% for training and 30% for testing), train , and test the model
   
    fitting = model.fit(predictors_norm, target, validation_split=0.3, epochs=50, verbose=0)
    #Find the last mean_squared_error in history.
    mean_squared_error = fitting.history['val_loss'][-1]
    #AAppend every mean_squared_error considered to the list.
    mean_squared_errors.append(mean_squared_error)
    print('i #{}: mean_squared_error {}'.format(i+1, mean_squared_error))

i #1: mean_squared_error 67.77114868164062
i #2: mean_squared_error 64.51091766357422
i #3: mean_squared_error 62.744537353515625
i #4: mean_squared_error 64.35626983642578
i #5: mean_squared_error 66.3448257446289
i #6: mean_squared_error 65.62712097167969
i #7: mean_squared_error 66.3621826171875
i #8: mean_squared_error 67.21324157714844
i #9: mean_squared_error 68.71736145019531
i #10: mean_squared_error 70.15755462646484
i #11: mean_squared_error 72.65748596191406
i #12: mean_squared_error 76.44063568115234
i #13: mean_squared_error 80.10762786865234
i #14: mean_squared_error 82.4648208618164
i #15: mean_squared_error 86.91581726074219
i #16: mean_squared_error 89.10697174072266
i #17: mean_squared_error 92.10164642333984
i #18: mean_squared_error 95.48860168457031
i #19: mean_squared_error 98.07377624511719
i #20: mean_squared_error 98.91725158691406
i #21: mean_squared_error 101.72447967529297
i #22: mean_squared_error 102.79324340820312
i #23: mean_squared_error 105.37487030029

Report the mean and the standard deviation of the mean squared errors.

In [21]:
print('The mean of the mean squared errors: {}'.format(np.mean(mean_squared_errors)))
print('The standard deviation of the mean squared errors: {}'.format(np.std(mean_squared_errors)))

The mean of the mean squared errors: 107.7179347229004
The standard deviation of the mean squared errors: 28.981621351744014


Investigate the effect of normalizing the predictors.

In [35]:
normalized_to_unnormalized= 107.7179347229004/110.9718231201172
normalized_to_unnormalized

0.9706782469123288

In [36]:
Effect_of_normalization= 1- (107.7179347229004/110.9718231201172)
Effect_of_normalization

0.029321753087671154

The mean of the mean squared errors with normalized predicors is almost 97 % of that without noramlized predictors. Thus, the effect of normalization is just about 3 %.